grpc_server.go

```diff
@@ -331,8 +337,36 @@ func (s *server) Subscribe(in *hegel.SubscribeRequest, stream hegel.Hegel_Subscr
 		}
 	}
 
+	s.subLock.RLock()
+	sub := s.subscriptions[id]
+	s.subLock.RUnlock()
+	// Disconnect previous client if a client is already connected for this hardware id
+	if sub != nil {
+		sub.cancel()
+	}
+	sub = &subscription{
+		Id:           id,
+		Ip:           ip,
+		StartedAt:    startedAt,
+		InitDuration: time.Since(startedAt),
+		cancel:       cancel,
+		updateChan:   make(chan []byte, 1),
+	}
+	s.subLock.Lock()
+	s.subscriptions[id] = sub
+	s.subLock.Unlock()
+
+	defer func() {
+		s.subLock.Lock()
+		defer s.subLock.Unlock()
+		if cSub, ok := s.subscriptions[id]; ok && cSub.updateChan == sub.updateChan {
```

Comment:
Don't need the `ok` check I think since for the `==` to be true then `ok` will also be true. If ok is false, there's no way `cSub.updateChan == sub.updateChan` because cSub.updateChan will be nil and `sub.updateChan` is never nil at this point.

Reply:
Also deserves some sort of comment that this is a guard against deleting a sub that has already been deleted because of a reconnection attempt or similar.

Reply:
`cSub` could be `null` if the id doesn't already exist. But will add a comment

Reply:
right, but `sub` isn't null so they don't `==` so you wont delete

Reply:
on a normal disconnect, cSub and sub could be equal. As another connection would not have come in and overwrite the `s.subscriptions[id]`